linera-sdk/src/lib.rs

```diff
@@ -265,71 +279,84 @@ pub struct QueryContext {
     /// Adds an effect to the execution result.
-    pub fn with_effect(
-        mut self,
-        destination: impl Into<Destination>,
-        effect: &impl Serialize,
-    ) -> Self {
-        let effect_bytes = bcs::to_bytes(effect).expect("Effect should be serializable");
-        self.effects.push((destination.into(), false, effect_bytes));
+    pub fn with_effect(mut self, destination: impl Into<Destination>, effect: Effect) -> Self {
+        self.effects.push((destination.into(), false, effect));
         self
     }
 
     /// Adds an authenticated effect to the execution result.
     pub fn with_authenticated_effect(
         mut self,
         destination: impl Into<Destination>,
-        effect: &impl Serialize,
+        effect: Effect,
     ) -> Self {
-        let effect_bytes = bcs::to_bytes(effect).expect("Effect should be serializable");
-        self.effects.push((destination.into(), true, effect_bytes));
+        self.effects.push((destination.into(), true, effect));
         self
     }
 }
 
 /// The result of calling into a user application.
-#[derive(Debug, Default, Deserialize, Serialize)]
+#[derive(Debug, Deserialize, Serialize)]
 #[cfg_attr(any(test, feature = "test"), derive(Eq, PartialEq))]
-pub struct ApplicationCallResult {
+pub struct ApplicationCallResult<Effect, Value, SessionType> {
     /// The return value.
-    #[debug(with = "linera_base::hex_debug")]
-    pub value: Vec<u8>,
+    // #[debug(with = "linera_base::hex_debug")]
+    pub value: Option<Value>,
     /// The externally-visible result.
-    pub execution_result: ExecutionResult,
+    pub execution_result: ExecutionResult<Effect>,
     /// The new sessions that were just created by the callee for us.
-    pub create_sessions: Vec<Session>,
+    pub create_sessions: Vec<Session<SessionType>>,
+}
+
+impl<Effect, Value, Session> Default for ApplicationCallResult<Effect, Value, Session> {
+    fn default() -> Self {
+        Self {
+            value: None,
+            execution_result: Default::default(),
+            create_sessions: vec![],
+        }
+    }
 }
 
 /// Syscall to request creating a new session.
 #[derive(Debug, Default, Deserialize, Serialize)]
 #[cfg_attr(any(test, feature = "test"), derive(Eq, PartialEq))]
-pub struct Session {
+pub struct Session<T> {
     /// A kind provided by the creator (meant to be visible to other applications).
     pub kind: u64,
     /// The data associated to the session.
-    #[debug(with = "linera_base::hex_debug")]
-    pub data: Vec<u8>,
+    pub data: T,
 }
 
 /// The result of calling into a session.
 #[derive(Default, Deserialize, Serialize)]
-pub struct SessionCallResult {
+pub struct SessionCallResult<Effect, Value, Session> {
     /// The application result.
-    pub inner: ApplicationCallResult,
+    pub inner: ApplicationCallResult<Effect, Value, Session>,
     /// If `call_session` was called, this tells the system to clean up the session.
     pub data: Option<Vec<u8>>,
```

Comment:
I believe this `Vec<u8>` is meant to be the session state (see above). `data` could be called `new_state`. (`None` means discard the object)